### PR TITLE
Reuse some abstract classes

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

File: src/library/scala/collection/SortedMap.scala [hunk 1]
```diff
@@ -114,7 +114,7 @@ trait SortedMapOps[K, +V, +CC[X, Y] <: Map[X, Y] with SortedMapOps[X, Y, CC, _],
   override def keySet: SortedSet[K] = new KeySortedSet
 
   /** The implementation class of the set returned by `keySet` */
-  protected class KeySortedSet extends SortedSet[K] with GenKeySet with GenKeySortedSet {
+  protected class KeySortedSet extends AbstractSet[K] with SortedSet[K] with GenKeySet with GenKeySortedSet {
```
Comment (joshlemer) [inline src/library/scala/collection/SortedMap.scala:117]: What's the actual convention here? Why AbstractSet rather than AbstractSortedSet?

Comment (dwijnand) [reply src/library/scala/collection/SortedMap.scala:117]: There is no `AbstractSortedSet`, or I'd've used it.  Not 100% sure what the convention is, but I can introduce and use (throughout) `AbstractSortedSet` if preferred.

Comment (NthPortal) [reply src/library/scala/collection/SortedMap.scala:117]: you can't actually introduce it unless this is making it into 2.13.0 (which unless I'm very much mistaken, it's not)

Comment (NthPortal) [reply src/library/scala/collection/SortedMap.scala:117]: in fact, checking the build output, even this change is not binary compatible

```scala
[error]  * the type hierarchy of class scala.<:< is different in other version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.$less$colon$less")
[error]  * the type hierarchy of class scala.=:= is different in other version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.$eq$colon$eq")
[error]  * the type hierarchy of class scala.collection.SortedMapOps#KeySortedSet is different in other version. Missing types {scala.collection.AbstractIterable,scala.collection.AbstractSet}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.collection.SortedMapOps$KeySortedSet")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaIntPredicate is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcZI$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaIntPredicate")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaDoubleConsumer is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcVD$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaDoubleConsumer")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaToIntFunction is different in other version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaToIntFunction")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaObjDoubleConsumer is different in other version. Missing types {scala.runtime.AbstractFunction2}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaObjDoubleConsumer")
[error]  * the type hierarchy of class scala.jdk.Accumulator#mcI#sp is different in other version. Missing types {scala.collection.mutable.AbstractSeq,scala.collection.AbstractIterable,scala.collection.AbstractSeq}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.Accumulator$mcI$sp")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaLongConsumer is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcVJ$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaLongConsumer")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaObjLongConsumer is different in other version. Missing types {scala.runtime.AbstractFunction2}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaObjLongConsumer")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaDoubleBinaryOperator is different in other version. Missing types {scala.runtime.AbstractFunction2,scala.runtime.AbstractFunction2$mcDDD$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaDoubleBinaryOperator")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaToLongBiFunction is different in other version. Missing types {scala.runtime.AbstractFunction2}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaToLongBiFunction")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaToDoubleFunction is different in other version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaToDoubleFunction")
[error]  * the type hierarchy of class scala.jdk.Accumulator#mcJ#sp is different in other version. Missing types {scala.collection.mutable.AbstractSeq,scala.collection.AbstractIterable,scala.collection.AbstractSeq}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.Accumulator$mcJ$sp")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaIntSupplier is different in other version. Missing types {scala.runtime.AbstractFunction0,scala.runtime.AbstractFunction0$mcI$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaIntSupplier")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaFunction is different in other version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaFunction")
[error]  * the type hierarchy of class scala.jdk.Accumulator#mcD#sp is different in other version. Missing types {scala.collection.mutable.AbstractSeq,scala.collection.AbstractIterable,scala.collection.AbstractSeq}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.Accumulator$mcD$sp")
[error]  * the type hierarchy of class scala.jdk.DoubleAccumulator is different in other version. Missing types {scala.collection.mutable.AbstractSeq,scala.collection.AbstractIterable,scala.collection.AbstractSeq}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.DoubleAccumulator")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaBooleanSupplier is different in other version. Missing types {scala.runtime.AbstractFunction0,scala.runtime.AbstractFunction0$mcZ$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaBooleanSupplier")
[error]  * the type hierarchy of class scala.jdk.AnyAccumulator is different in other version. Missing types {scala.collection.mutable.AbstractSeq,scala.collection.AbstractIterable,scala.collection.AbstractSeq}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.AnyAccumulator")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaDoubleSupplier is different in other version. Missing types {scala.runtime.AbstractFunction0,scala.runtime.AbstractFunction0$mcD$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaDoubleSupplier")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaUnaryOperator is different in other version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaUnaryOperator")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaLongSupplier is different in other version. Missing types {scala.runtime.AbstractFunction0,scala.runtime.AbstractFunction0$mcJ$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaLongSupplier")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaIntToDoubleFunction is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcDI$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaIntToDoubleFunction")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaDoubleUnaryOperator is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcDD$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaDoubleUnaryOperator")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaDoubleToLongFunction is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcJD$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaDoubleToLongFunction")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaBiPredicate is different in other version. Missing types {scala.runtime.AbstractFunction2}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaBiPredicate")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaIntBinaryOperator is different in other version. Missing types {scala.runtime.AbstractFunction2,scala.runtime.AbstractFunction2$mcIII$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaIntBinaryOperator")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaDoubleFunction is different in other version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaDoubleFunction")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaLongToIntFunction is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcIJ$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaLongToIntFunction")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaConsumer is different in other version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaConsumer")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaLongUnaryOperator is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcJJ$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaLongUnaryOperator")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaLongBinaryOperator is different in other version. Missing types {scala.runtime.AbstractFunction2,scala.runtime.AbstractFunction2$mcJJJ$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaLongBinaryOperator")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaBiFunction is different in other version. Missing types {scala.runtime.AbstractFunction2}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaBiFunction")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaLongFunction is different in other version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaLongFunction")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaIntUnaryOperator is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcII$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaIntUnaryOperator")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaPredicate is different in other version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaPredicate")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaIntToLongFunction is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcJI$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaIntToLongFunction")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaBiConsumer is different in other version. Missing types {scala.runtime.AbstractFunction2}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaBiConsumer")
[error]  * the type hierarchy of class scala.jdk.LongAccumulator is different in other version. Missing types {scala.collection.mutable.AbstractSeq,scala.collection.AbstractIterable,scala.collection.AbstractSeq}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.LongAccumulator")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaSupplier is different in other version. Missing types {scala.runtime.AbstractFunction0}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaSupplier")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaDoublePredicate is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcZD$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaDoublePredicate")
[error]  * the type hierarchy of class scala.jdk.IntAccumulator is different in other version. Missing types {scala.collection.mutable.AbstractSeq,scala.collection.AbstractIterable,scala.collection.AbstractSeq}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.IntAccumulator")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaIntFunction is different in other version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaIntFunction")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaBinaryOperator is different in other version. Missing types {scala.runtime.AbstractFunction2}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaBinaryOperator")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaLongPredicate is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcZJ$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaLongPredicate")
[error]  * the type hierarchy of class scala.jdk.Accumulator is different in other version. Missing types {scala.collection.mutable.AbstractSeq,scala.collection.AbstractIterable,scala.collection.AbstractSeq}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.Accumulator")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaIntConsumer is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcVI$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaIntConsumer")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaDoubleToIntFunction is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcID$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaDoubleToIntFunction")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaLongToDoubleFunction is different in other version. Missing types {scala.runtime.AbstractFunction1,scala.runtime.AbstractFunction1$mcDJ$sp}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaLongToDoubleFunction")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaToIntBiFunction is different in other version. Missing types {scala.runtime.AbstractFunction2}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaToIntBiFunction")
[error]  * the type hierarchy of class scala.jdk.FunctionWrappers#FromJavaToLongFunction is different in other version. Missing types {scala.runtime.AbstractFunction1}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("scala.jdk.FunctionWrappers$FromJavaToLongFunction")
```

Comment (dwijnand) [reply src/library/scala/collection/SortedMap.scala:117]: Yup, it's forwards-incompatible, unless we're happy with, for instance, `scala.<:<` having a `scala.runtime.AbstractFunction1` parent in 2.13.1.  (`<:<` is sealed, so particularly unproblematic, but the others aren't).

I think `MissingTypesProblem` is a problem if it imposes abstract definitions downstream, such as having to implement a method.  There is no such imposition, so it should be fine, no?  (I'm unsure.)

Comment (sjrd) [reply src/library/scala/collection/SortedMap.scala:117]: No it's not fine, due to type inference in lubbing. Such a change was erroneously thought she in 2.11.9 or 2.11.10, and was the reason that one of those two versions was blacklisted.

You can't do those changes in a minor Scala release.

In addition, I think it's really bad anyway to reference `runtime` in source code, especially in what amounts to public API. `runtime` is an implementation detail of the language and compiler. It should even be hidden from the rest of the standard library.

**Comment by dwijnand** [comment]
I fast-tracked this change out of some more long-term work I was doing because I understood the use of "AbstractX" (whether in scala.runtime or not) was to reduce the amount of bytecode in scala-library.

I take it that isn't so hotly desired or, more likely, the gain here is negligible.

No need to keep this hanging around until 2.14 days.

**Pull request closed without merging**